pubnub/models/subscription.py

```diff
@@ -0,0 +1,338 @@
+    on_message: callable
+    on_signal: callable
+    on_presence: callable
+    on_channel_metadata: callable
+    on_user_metadata: callable
+    on_message_action: callable
+    on_membership: callable
+    on_file: callable
+
+    def is_matching_listener(self, message):
+        def wildcard_match(name, subscription):
+            return subscription.endswith('.*') and name.startswith(subscription.strip('*'))
+
+        if self.subscription_type == PNSubscriptionType.CHANNEL:
+            return any(
+                [message.channel == subscription or wildcard_match(message.channel, subscription)
+                    for subscription in self.subscription_names]
+            )
+        elif self.subscription_type == PNSubscriptionType.CHANNEL_GROUP:
+            return message.subscription in self.subscription_names
+        else:
+            return any(
+                [message.channel == subscription or wildcard_match(message.channel, subscription)
+                    for subscription in (self.channels + self.channel_groups)]
+            )
+
+    def presence(self, presence):
+        if not hasattr(self, 'on_presence') or not self.with_presence:
+            return
+
+        if self.is_matching_listener(presence) and hasattr(self, 'on_presence'):
+            self.on_presence(presence)
+
+    def message(self, message):
+        if self.is_matching_listener(message) and hasattr(self, 'on_message'):
+            self.on_message(message)
+
+    def message_action(self, message_action):
+        if self.is_matching_listener(message_action) and hasattr(self, 'on_message_action'):
+            self.on_message_action(message_action)
+
+    def signal(self, signal):
+        if self.is_matching_listener(signal) and hasattr(self, 'on_signal'):
+            self.on_signal(signal)
+
+
+class PNSubscribeCapable:
+    def subscribe(self):
+        self.subscription_registry.add(self)
+
+    def unsubscribe(self):
+        self.subscription_registry.remove(self)
+
+
+class PubNubSubscription(PNEventEmitter, PNSubscribeCapable):
+    def __init__(self, pubnub_instance,
+                 subscription_names: List[str],
+                 subscription_type: PNSubscriptionType,
+                 timetoken: Optional[int] = None,
+                 region: Optional[str] = None,
+                 with_presence: bool = False) -> None:
+
+        self.subscription_registry = pubnub_instance._subscription_registry
+        self.subscription_manager = pubnub_instance._subscription_manager
+        self.subscription = None
+        self.subscription_names = subscription_names
+        self.subscription_type = subscription_type
+        self.timetoken = timetoken
+        self.region = region
+        self.with_presence = with_presence
+
+    def add_listener(self, listener):
+        self.subscription_registry.add_listener(listener)
+
+
+class PubNubSubscriptionSet(PNEventEmitter, PNSubscribeCapable):
+    def __init__(self, pubnub_instance,
```

Comment:
how do I create a subscriptionset from existing subscriptions?
also how do I add/remove subscriptions to a set?

looking at the sset class it doesn't seem to hold subscriptions at all, just a list of channel/group names?
It was supposed to be a class that lets you act on multiple Subscription objects

Reply:
for example, I should be able to add Subscription objects to a Set, then if I call set.subscribe() or set.unsubscribe() then all those subscriptions should be subscribed/unsubscribed. I don't suppose I can do that here?